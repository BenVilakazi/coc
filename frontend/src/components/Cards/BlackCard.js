import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import Markdown from 'markdown-to-jsx';
import LayeredCards from './LayeredCards';

const propTypes = {
  pickCount: PropTypes.number.isRequired,
  children: PropTypes.string,
  winnerScreen: PropTypes.bool,
};

const defaultProps = {
  children: '',
  winnerScreen: false,
};

const BLANK_LENGTH = 8;

const StyledBlackCard = styled.div`
  position: relative;
  width: 390px;
  height: 540px;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
  border-radius: 18px;
  font-weight: 700;
  font-size: 36px;
  padding: 45px 21px 0 21px;
  box-shadow: 4px 4px 24px rgba(0, 0, 0, 0.3);

  ${(props) =>
    props.winnerDisplay &&
    css`
      width: 276px;
      height: 363px;
      font-size: 24px;
    `}

  .pickText {
    font-size: 24px;
    margin-right: 7px;
  }

  .layeredCards {
    position: absolute;
    bottom: 18px;
    left: 18px;
  }

  .pickCount {
    position: absolute;
    bottom: 8px;
    right: 21px;
    display: flex;
    justify-content: center;
    align-items: center;
  }
`;

const Circled = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  color: var(--primary-text-color);
  background: var(--primary-background-color);
  width: 20px;
  height: 20px;
  border-radius: 50%;
  font-weight: bold;
  font-size: 14px;
`;

function parseForMarkdown(string, blankLength) {
  return (
    string
      // https://regexr.com/5ltnd
      // This regex finds all instances of the newline character \n
      // Then it gets replaced with the markdown equivalent \n\n
      .replace(/\n/g, '\n\n')

      // https://regexr.com/5td92
      // These regex find all underscores used for markdown purposes
      // Then it replaces them with their asterisk variant
      .replace(/(___)([^\s_]+)(___)/g, `***$2***`)
      .replace(/(__)([^\s_]+)(__)/g, `**$2**`)
      .replace(/(_)([^\s_]+)(_)/g, `*$2*`)

      // https://regexr.com/5ltng
      // This regex finds all remaining instances of underscores
      // Then it gets replaced with a long blank line (multiple escaped underscores)
      .replace(/_/g, `${'\\_'.repeat(blankLength)}`)
  );
}

function BlackCard({ pickCount, children, winnerScreen }) {
  return (
    <StyledBlackCard data-testid="black-card" winnerDisplay={winnerScreen}>
      <Markdown options={{ wrapper: 'div' }}>
        {parseForMarkdown(children, BLANK_LENGTH)}
      </Markdown>
      <div className="layeredCards">
        <LayeredCards />
      </div>
      <div className="pickCount">
        <span className="pickText">PICK</span>
        <Circled>{pickCount}</Circled>
      </div>
    </StyledBlackCard>
  );
}

BlackCard.propTypes = propTypes;

BlackCard.defaultProps = defaultProps;

export default BlackCard;
