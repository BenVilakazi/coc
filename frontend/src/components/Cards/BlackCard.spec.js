import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import BlackCard from './BlackCard';

describe('BlackCard', () => {
  describe('rendering', () => {
    it('renders the BlackCard component given a pickCount number and a string', () => {
      render(<BlackCard pickCount={0}>Howdy</BlackCard>);

      expect(screen.getByText('Howdy')).toBeInTheDocument();
      expect(screen.getByText('PICK')).toBeInTheDocument();
      expect(screen.getByText('0')).toBeInTheDocument();

      // Two SVGs in the bottom left of the card
      expect(screen.queryAllByTestId('layerCard')).toHaveLength(1);
    });

    it('renders a black card with no main text present', () => {
      render(<BlackCard pickCount={0} />);
      expect(screen.getByTestId('black-card')).toBeInTheDocument();
      expect(screen.getByText('PICK')).toBeInTheDocument();
      expect(screen.getByText('0')).toBeInTheDocument();

      // Two SVGs in the bottom left of the card
      expect(screen.queryAllByTestId('layerCard')).toHaveLength(1);
    });
  });

  describe('text parsing', () => {
    it('properly converts markdown to JSX', () => {
      const string = 'This _is_ a **test**\n\n Here is a *new* __line__';
      const tree = renderer
        .create(<BlackCard pickCount={0}>{string}</BlackCard>)
        .toJSON();

      expect(screen.queryAllByText('\n')).toHaveLength(0);
      expect(tree).toMatchSnapshot();
    });

    it('does not display markdown', () => {
      render(<BlackCard pickCount={0}># **This** _is_ a test</BlackCard>);

      expect(screen.queryByText('#')).not.toBeInTheDocument();
      expect(screen.queryByText('**')).not.toBeInTheDocument();
      expect(screen.queryByText('_')).not.toBeInTheDocument();
    });

    it('properly converts single underscores to longer blanks', () => {
      render(<BlackCard pickCount={0}>This is a _ test.</BlackCard>);

      expect(screen.getByText(/________/)).toBeInTheDocument();
    });

    it('properly converts single underscores next to punctuation to longer blanks', () => {
      render(<BlackCard pickCount={0}>This is a _.</BlackCard>);

      expect(screen.getByText(/________./)).toBeInTheDocument();
    });

    it('properly converts single underscores between the beginning of text and punctuation to longer blanks', () => {
      render(<BlackCard pickCount={0}>_! This is a test.</BlackCard>);

      expect(screen.getByText(/________./)).toBeInTheDocument();
    });

    it('properly converts single underscores between quotes to longer blanks', () => {
      render(
        <BlackCard pickCount={0}> This is a &quot;_&quot; test.</BlackCard>,
      );

      expect(screen.getByText(/________./)).toBeInTheDocument();
    });

    it('properly converts underscores at the beginning of a word into longer blanks', () => {
      render(<BlackCard pickCount={0}> This _is a test.</BlackCard>);

      expect(screen.getByText(/________./)).toBeInTheDocument();
    });

    it('properly converts underscores at the end of a word into longer blanks', () => {
      render(<BlackCard pickCount={0}> This is_ a test.</BlackCard>);

      expect(screen.getByText(/________./)).toBeInTheDocument();
    });

    it('does not display single underscores by themselves', () => {
      render(<BlackCard pickCount={0}>This _ should be long</BlackCard>);

      expect(screen.queryByText(/ _ /)).not.toBeInTheDocument();
    });
  });
});
