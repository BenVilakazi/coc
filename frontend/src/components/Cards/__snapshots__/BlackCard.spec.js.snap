// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlackCard text parsing properly converts markdown to JSX 1`] = `
.c0 {
  position: relative;
  width: 390px;
  height: 540px;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
  border-radius: 18px;
  font-weight: 700;
  font-size: 36px;
  padding: 45px 21px 0 21px;
  box-shadow: 4px 4px 24px rgba(0,0,0,0.3);
}

.c0 .pickText {
  font-size: 24px;
  margin-right: 7px;
}

.c0 .layeredCards {
  position: absolute;
  bottom: 18px;
  left: 18px;
}

.c0 .pickCount {
  position: absolute;
  bottom: 8px;
  right: 21px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: var(--primary-text-color);
  background: var(--primary-background-color);
  width: 20px;
  height: 20px;
  border-radius: 50%;
  font-weight: bold;
  font-size: 14px;
}

<div
  className="c0"
  data-testid="black-card"
>
  <div>
    <p>
      This 
      <em>
        is
      </em>
       a 
      <strong>
        test
      </strong>
    </p>
    <p>
       Here is a 
      <em>
        new
      </em>
       
      <strong>
        line
      </strong>
    </p>
  </div>
  <div
    className="layeredCards"
  >
    <div
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <img
        alt="offset black and white card stack"
        data-testid="layerCard"
        src="cards-stack-offset-dark-icon.svg"
      />
    </div>
  </div>
  <div
    className="pickCount"
  >
    <span
      className="pickText"
    >
      PICK
    </span>
    <div
      className="c1"
    >
      0
    </div>
  </div>
</div>
`;
